ktor-client/ktor-client-curl/posix/src/io/ktor/client/engine/curl/CurlProcessor.kt

```diff
@@ -20,72 +18,45 @@ private lateinit var curlApi: CurlMultiApiHandler
 internal class CurlProcessor(
     override val coroutineContext: CoroutineContext
 ) : CoroutineScope {
-    private val worker: Worker = Worker.start()
-    private val responseConsumers: ConcurrentMap<CurlRequestData, CompletableDeferred<CurlSuccess>> = ConcurrentMap()
-    private val activeRequests = atomic(0)
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private val curlDispatcher: CoroutineDispatcher = newSingleThreadContext("curl-dispatcher")
+    private val curlScope = CoroutineScope(coroutineContext + curlDispatcher)
 
     init {
-        worker.execute(TransferMode.SAFE, { Unit }) {
+        curlScope.launch {
             curlApi = CurlMultiApiHandler()
         }
     }
 
-    public suspend fun executeRequest(request: CurlRequestData, callContext: CoroutineContext): CurlSuccess {
+    public suspend fun executeRequest(request: CurlRequestData): CurlSuccess {
         val deferred = CompletableDeferred<CurlSuccess>()
-        responseConsumers[request] = deferred
 
-        val easyHandle = worker.execute(TransferMode.SAFE, { request.freeze() }, ::curlSchedule).result
+        request.freeze()
+        val easyHandle = withContext(curlScope.coroutineContext) { curlApi.scheduleRequest(request, deferred) }
 
-        val requestCleaner = callContext[Job]!!.invokeOnCompletion { cause ->
+        val requestCleaner = request.executionContext.invokeOnCompletion { cause ->
             if (cause == null) return@invokeOnCompletion
             cancelRequest(easyHandle, cause)
         }
 
         try {
-            activeRequests.incrementAndGet()
-
-            while (deferred.isActive) {
-                val completedResponses = poll()
-                while (completedResponses.state == FutureState.SCHEDULED) delay(100)
-                processPoll(completedResponses.result)
-            }
-
+            curlPerform()
             return deferred.await()
         } finally {
             requestCleaner.dispose()
         }
     }
 
     public fun close() {
-        worker.execute(TransferMode.SAFE, { Unit }) { curlApi.close() }
-        worker.requestTermination()
+        curlScope.launch { curlApi.close() }
+        curlScope.cancel()
```

Comment:
it looks like `launch` and `cancel` are in the race.

Reply:
thanks, updated